Tests pick up configuration file in default location

While running the framework test suite, I experienced failing tests since the configuration file `$HOME/.config/easybuild/config.cfg` was picked up and evaluated.

In my particular case, setting `sourcepath` in `$HOME/.config/easybuild/config.cfg` led to three failures:
 * test.framework.config
```
======================================================================
FAIL: test_default_config (__main__.EasyBuildConfigTest)
Test default configuration.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/geimer/Projects/EasyBuild/easybuild-framework/test/framework/config.py", line 119, in test_default_config
    self.assertEqual(config_options['sourcepath'], os.path.join(eb_homedir, 'sources'))
AssertionError: '/home/geimer/Projects/EasyBuild/Archives/Sources' != '/home/geimer/.local/easybuild/sources'

======================================================================
FAIL: test_generaloption_config_file (__main__.EasyBuildConfigTest)
Test use of new-style configuration file.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/geimer/Projects/EasyBuild/easybuild-framework/test/framework/config.py", line 277, in test_generaloption_config_file
    self.assertEqual(source_paths(), [os.path.join(os.getenv('HOME'), '.local', 'easybuild', 'sources')])  # default
AssertionError: Lists differ: ['/home/geimer/Projects/EasyBu... != ['/home/geimer/.local/easybuil...

First differing element 0:
/home/geimer/Projects/EasyBuild/Archives/Sources
/home/geimer/.local/easybuild/sources

- ['/home/geimer/Projects/EasyBuild/Archives/Sources']
+ ['/home/geimer/.local/easybuild/sources']

----------------------------------------------------------------------
```
 * test.framework.options
```
FAIL: test_generate_cmd_line (__main__.CommandLineOptionsTest)
Test for generate_cmd_line.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/geimer/Projects/EasyBuild/easybuild-framework/test/framework/options.py", line 1674, in test_generate_cmd_line
    self.assertEqual(ebopts.generate_cmd_line(), [])
AssertionError: Lists differ: ['--sourcepath=/home/geimer/Pr... != []

First list contains 1 additional elements.
First extra element 0:
--sourcepath=/home/geimer/Projects/EasyBuild/Archives/Sources

- ['--sourcepath=/home/geimer/Projects/EasyBuild/Archives/Sources']
+ []

----------------------------------------------------------------------
```
@boegel and I discussed this issue already at the Hackathon last week and came to the conclusion that this should be handled in the framework, as other settings may (and in fact do) affect other tests.

--- a/test/framework/config.py
+++ b/test/framework/config.py
@@ -80,7 +80,8 @@ class EasyBuildConfigTest(EnhancedTestCase):
     def purge_environment(self):
         """Remove any leftover easybuild variables"""
         for var in os.environ.keys():
-            if var.startswith('EASYBUILD_'):
+            # retain $EASYBUILD_IGNORECONFIGFILES, to make sure the test is isolated from system-wide config files!
+            if var.startswith('EASYBUILD_') and var != 'EASYBUILD_IGNORECONFIGFILES':
                 del os.environ[var]
 
     def tearDown(self):
@@ -425,7 +426,10 @@ class EasyBuildConfigTest(EnhancedTestCase):
         mkdir(os.path.join(dir3, 'easybuild.d'), parents=True)
         write_file(os.path.join(dir3, 'easybuild.d', 'foobarbaz.cfg'), cfg_template % '/foobarbaz')
 
-        # only $XDG_CONFIG_HOME set
+        # set $XDG_CONFIG_DIRS to non-existing dir to isolate ourselves from possible system-wide config files
+        os.environ['XDG_CONFIG_DIRS'] = '/there/should/be/no/such/directory/we/hope'
+
+        # only $XDG_CONFIG_HOME set (to existing path)
         os.environ['XDG_CONFIG_HOME'] = homedir
         cfg_files = [os.path.join(homedir, 'easybuild', 'config.cfg')]
         reload(eboptions)

--- a/test/framework/options.py
+++ b/test/framework/options.py
@@ -72,7 +72,8 @@ class CommandLineOptionsTest(EnhancedTestCase):
     def purge_environment(self):
         """Remove any leftover easybuild variables"""
         for var in os.environ.keys():
-            if var.startswith('EASYBUILD_'):
+            # retain $EASYBUILD_IGNORECONFIGFILES, to make sure the test is isolated from system-wide config files!
+            if var.startswith('EASYBUILD_') and var != 'EASYBUILD_IGNORECONFIGFILES':
                 del os.environ[var]
 
     def test_help_short(self, txt=None):

--- a/test/framework/options.py
+++ b/test/framework/options.py
@@ -1671,14 +1671,24 @@ class CommandLineOptionsTest(EnhancedTestCase):
 
     def test_generate_cmd_line(self):
         """Test for generate_cmd_line."""
-        ebopts = EasyBuildOptions()
-        self.assertEqual(ebopts.generate_cmd_line(), [])
+        self.purge_environment()
 
-        ebopts = EasyBuildOptions(go_args=['--force'])
-        self.assertEqual(ebopts.generate_cmd_line(), ['--force'])
+        def generate_cmd_line(ebopts):
+            """Helper function to filter generated command line (to ignore $EASYBUILD_IGNORECONFIGFILES)."""
+            return [x for x in ebopts.generate_cmd_line() if not x.startswith('--ignoreconfigfiles=')]
 
-        ebopts = EasyBuildOptions(go_args=['--search=bar', '--search', 'foobar'])
-        self.assertEqual(ebopts.generate_cmd_line(), ['--search=foobar'])
+        ebopts = EasyBuildOptions(envvar_prefix='EASYBUILD')
+        self.assertEqual(generate_cmd_line(ebopts), [])
+
+        ebopts = EasyBuildOptions(go_args=['--force'], envvar_prefix='EASYBUILD')
+        self.assertEqual(generate_cmd_line(ebopts), ['--force'])
+
+        ebopts = EasyBuildOptions(go_args=['--search=bar', '--search', 'foobar'], envvar_prefix='EASYBUILD')
+        self.assertEqual(generate_cmd_line(ebopts), ['--search=foobar'])
+
+        os.environ['EASYBUILD_DEBUG'] = '1'
+        ebopts = EasyBuildOptions(go_args=['--force'], envvar_prefix='EASYBUILD')
+        self.assertEqual(generate_cmd_line(ebopts), ['--debug', '--force'])
 
     def test_include_easyblocks(self):
         """Test --include-easyblocks."""

